testing/sync-test/src/auth.rs

```diff
@@ -253,3 +273,102 @@ pub fn cleanup_server(clients: Vec<&mut TestClient>) -> Result<(), failure::Erro
     }
     failure::bail!("None of the clients managed to wipe the server!");
 }
+
+pub struct TestUser {
+    pub account: Arc<TestAccount>,
+    pub clients: Vec<TestClient>,
+}
+
+impl TestUser {
+    fn new_random(opts: &Opts, client_count: usize) -> Result<Self, failure::Error> {
+        log::info!("Creating test account with {} clients", client_count);
+
+        let account = TestAccount::new_random(&opts)?;
+        let mut clients = Vec::with_capacity(client_count);
+
+        for c in 0..client_count {
+            log::info!("Creating test client {}", c);
+            clients.push(TestClient::new(account.clone())?);
+        }
+        Ok(Self { account, clients })
+    }
+
+    pub fn new(opts: &Opts, client_count: usize) -> Result<TestUser, failure::Error> {
+        if opts.oauth_retries > 0 && opts.no_delete_account {
+            failure::bail!(
+                "Illegal option combination: oauth-retries is nonzero \
+                 and no-delete-account is specified."
+            );
+        }
+        if opts.helper_debug {
+            std::env::set_var("DEBUG", "nightmare");
+            std::env::set_var("HELPER_SHOW_BROWSER", "1");
+        }
+        for attempt in 0..=opts.oauth_retries {
+            log::info!("Creating test user (attempt {})", attempt);
+            match TestUser::new_random(opts, client_count) {
+                Ok(user) => {
+                    log::info!("Created test user (attempt {})", attempt);
+                    return Ok(user);
+                }
+                Err(e) => {
+                    if attempt == opts.oauth_retries {
+                        log::error!("Failed to create test user (attempt {}): {:?}", attempt, e);
+                        return Err(e);
+                    }
+                    log::warn!("Failed to create test user (attempt {}): {}", attempt, e);
+                    if opts.oauth_delay_time > 0 {
+                        let delay = opts.oauth_delay_time + attempt * opts.oauth_retry_backoff;
+                        log::info!(
+                            "Retrying after {} ms (attempt {} => {})",
+                            delay,
+                            attempt,
+                            attempt + 1
+                        );
+                        std::thread::sleep(std::time::Duration::from_millis(delay as u64));
+                    }
+                }
+            }
+        }
+        // Above loop always either hits the `return Err(e)` or `return Ok(user);` cases
+        unreachable!();
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum FxaConfigUrl {
```

Comment:
The intro of this type makes me think our `FxAConfig`/`FirefoxAccount` API could be better?

Reply:
Eh, it's just there to help structopt parse things.